rust/relay/ebpf-turn-router/src/main.rs

```diff
@@ -164,135 +164,661 @@ fn try_handle_turn_ipv4(ctx: &XdpContext, eth: Eth) -> Result<(), Error> {
-        3478,
-        client_and_channel.client_port(),
-        new_udp_len,
-        channel_number,
-        channel_data_length,
+    let key = ClientAndChannelV4::new(ipv4.src_addr(), udp.source(), channel_number);
+
+    // SAFETY: We only write to these using a single thread in userspace.
+    if let Some(port_and_peer) = unsafe { CHAN_TO_UDP_44.get(&key) } {
+        // IPv4 to IPv4 - existing logic
+        handle_ipv4_to_ipv4_channel_to_udp(ctx, port_and_peer)?;
+        return Ok(());
+    }
+
+    // SAFETY: We only write to these using a single thread in userspace.
+    if let Some(port_and_peer) = unsafe { CHAN_TO_UDP_46.get(&key) } {
+        handle_ipv4_to_ipv6_channel_to_udp(ctx, port_and_peer)?;
+        return Ok(());
+    }
+
+    Err(Error::NoEntry(SupportedChannel::Chan4ToUdp))
+}
+
+#[inline(always)]
+fn handle_ipv4_to_ipv4_channel_to_udp(
+    ctx: &XdpContext,
+    port_and_peer: &PortAndPeerV4,
+) -> Result<(), Error> {
+    const NET_SHRINK: i32 = CdHdr::LEN as i32; // Shrink by 4 bytes for channel data header
+
+    // SAFETY: XDP context provides valid packet memory. Bounds checked by ref_mut_at.
+    let old_eth = unsafe { ref_mut_at::<EthHdr>(ctx, 0)? };
+    let old_eth_src = old_eth.src_addr;
+    let old_eth_dst = old_eth.dst_addr;
+    let old_eth_type = old_eth.ether_type;
+
+    // SAFETY: XDP context provides valid packet memory. Bounds checked by ref_mut_at.
+    let old_ipv4 = unsafe { ref_mut_at::<Ipv4Hdr>(ctx, EthHdr::LEN)? };
+    let old_ipv4_src = old_ipv4.src_addr();
+    let old_ipv4_dst = old_ipv4.dst_addr();
+    let old_ipv4_len = old_ipv4.total_len();
+    let old_ipv4_check = old_ipv4.checksum();
+    let old_ipv4_tos = old_ipv4.tos;
+    let old_ipv4_id = old_ipv4.id();
+    let old_ipv4_frag_off = old_ipv4.frag_off;
+    let old_ipv4_ttl = old_ipv4.ttl;
+    let old_ipv4_proto = old_ipv4.proto;
+
+    // SAFETY: XDP context provides valid packet memory. Bounds checked by ref_mut_at.
+    let old_udp = unsafe { ref_mut_at::<UdpHdr>(ctx, EthHdr::LEN + Ipv4Hdr::LEN)? };
+    let old_udp_src = old_udp.source();
+    let old_udp_dst = old_udp.dest();
+    let old_udp_len = old_udp.len();
+    let old_udp_check = old_udp.check();
+
+    // Refuse to compute full UDP checksum.
+    // We forged these packets, so something's wrong if this is zero.
+    if old_udp_check == 0 {
+        return Err(Error::Ipv4ChecksumMissing);
+    }
+
+    // Read channel data header before we lose it
+    // SAFETY: XDP context provides valid packet memory. Bounds checked by ref_mut_at.
+    let old_cd = unsafe { ref_mut_at::<CdHdr>(ctx, EthHdr::LEN + Ipv4Hdr::LEN + UdpHdr::LEN)? };
+    let channel_number = u16::from_be_bytes(old_cd.number);
+    let channel_data_length = u16::from_be_bytes(old_cd.length);
+
+    //
+    // 1. Ethernet header
+    //
+    // SAFETY: After packet shrinking, the offset is valid within adjusted bounds.
+    let eth = unsafe { ref_mut_at::<EthHdr>(ctx, NET_SHRINK as usize)? };
```

Comment:
This is not safe and might trigger UB, you should make sure `old_eth` is out of scope before you call this.

Reply:
Alternatively, what would also work is a pattern like this:

```rust
let (a, b, c) = {
    let ipv4 = unsafe { ref_mut_at ... };

    let a = ipv4.src;
    // ...

   (a, b, c)
};
```

Basically, like a function call but just an inline scope.

You could also make functions to "read" variables from a memory section and return them as a tuple. Whatever makes the mutable reference as short-lived as possible.

I'd think that if we mark it as `#[inline(always)]` it should be fine?

Reply:
Hmm let me think about this some more. I don't know if it makes sense to follow borrow checking rules here if we are going to move bytes directly from one packet region to another.

Taking a non-mutable ref and then return its values so that they persist after the ref is out of scope would necessarily copying the variables onto the stack, which is what we do now but want to move away from.

Any ideas how we can model the header field-by-field copies using the convenient `network_types` structs? If not, I may have to go one level deeper (sorry) and do a manual field-by-field copy. It won't be that bad if I name the dereferences well and leave comments I think. But it will save copying to stack and back for fields that simply move their place.

Reply:
The entire header always moves by at least 4 bytes, so you do need to copy every field right? What more is there to optimise?

You don't have to use mutable references, you can just use pointers instead. All access will be unsafe then so it will be harder to audit.

Reply:
I am all for optimizations but this sounds premature to me. What tells you that loading a larger chunk of memory at once onto the stack isn't more efficient than individual copies?

Reply:
Have you considered moving the entire header first and then updating it in-place? Guessing performance is difficult but it seems like that is something the compiler should be able to generate pretty efficient code for because it is very predictable.

Reply:
> that it is not violating LLVM's aliasing rules.

(It is my understanding that mutable refs in Rust translate to specific LLVM IR such that it can reason about what optimisations it can make)

Reply:
After I saw how much performance improved by moving from u8 copies in a tight loop to u32 copies in a tight loop, my confidence in LLVM being able to generate decent output for the BPF target fell by quite a lot. This makes sense I suppose because it is trying to generate code that will verify, above all else. But I can gather some data to support any further optimizations.

Reply:
> Agreed. So I'll go with the inline scope solution?

Sounds good to me!

Reply:
> We will probably need a different approach to testing here anyway because we'll have a very hard time saturating performance with "live" clients and gateways.

It shouldn't be too difficult to write a small TURN client bundled with a TUN device so we can run iperf through it? Although I am not sure there is much value in it at this point. We want to optimise Firezone throughput so measuring that seems to make the most sense outside of having numbers to present for the blogpost.

We could maybe also do the "wrapping" in eBPF on the client / gw side?